### ISSUE TITLE
False-positive error due to missing and wrong type inference (involving IntoIterator, as cast)

### ISSUE THREAD

**Issue by feldroop** [rust-lang/rust-analyzer#18282]
**rust-analyzer version**: rust-analyzer version: 0.3.2137-standalone

**rustc version**: rustc 1.81.0

**editor or extension**: VSCode

**relevant settings**: -

**repository link (if public, optional)**: -

**code snippet to reproduce**: 
```rust 
fn problem<II, I>(iter: II)
where
    II: IntoIterator<Item = I>,
    I: IntoIterator<Item = ()>,
{
    for inner_iter in iter.into_iter() {
        for (index, _) in inner_iter.into_iter().enumerate() {
            index as f64;
            let _ = (1usize + index) as f64;
        }
    }
}
```

I believe this is connected to the type inference issue of https://github.com/rust-lang/rust-analyzer/issues/10653.

Additionally, I noticed that the no-op `index as f64` leads to `index` being inferred as an `f64`. This in turn leads to the next line generating the following false-positive error: **non-primitive cast: `<usize as Add<f64>>::Output` as `f64`**

**Comment by lnicola** [comment]
Yeah, this is the same as #10653.
